File: leetcode/easy/1022-sum-binary/sol2.cpp
/**
 * @file sol2.cpp
 * @author Vortexx2
 * @brief Problem 1022 - Sum Of Root to Leaf Binary
 * @approach Adding current path binary value to the answer when encountering a leaf node, rather than making 2 passes like sol.cpp.
 * @date 13-08-2021
 * 
 * Runtime - 3 ms
 * Memory Usage - 16.7 MB
 */
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  int ans;
  int sumRootToLeaf(TreeNode *root)
  {
    ans = 0;
    writePaths(root, 0);

    return ans;
  }

  void writePaths(TreeNode *root, int running)
  {
    if (!root)
      return;

    running = running << 1 | root->val;
    if (!root->left && !root->right)
      ans += running;

    else
    {
      writePaths(root->left, running);
      writePaths(root->right, running);
    }
  }
};

File: leetcode/easy/1022-sum-binary/sol2.py
"""
  @author Vortexx2
  Problem 1022

  Runtime - 43 ms
  Memory Usage - 14.3 MB
"""
from typing import Optional, List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:

    self.ans = 0
    self.writePaths(root, 0)

    return self.ans

  def writePaths(self, root: TreeNode, running: int) -> None:

    if not root:
      return

    running = running << 1 | root.val

    if not root.left and not root.right:
      self.ans += running

    else:
      self.writePaths(root.left, running)
      self.writePaths(root.right, running)
